forte/data/data_store.py

```diff
@@ -0,0 +1,239 @@
+        """
+
+        tid: int = uuid.uuid4().int
+        entry = [type, tid, begin, end]
+        entry += len(self._type_attributes[type]) * [None]
+        return entry
+    
+    def _generate_attr_index(self):
+        """
+        For every type in type_attributes, we need to convert the attribute list to
+        a dictionary, mapping each attribute string to a unique index.
+        The index should be the actual index of this attribute in the entry's list.
+        Index 0, 1, 2, and 3 should be reserved for begin, end, tid and type.
+
+        For example, for type `ft.onto.base_ontology.Sentence`, the attribute ``speaker`` 
+        has index 4, and attribute ``part_id`` has index 5.
+
+        """
+        raise NotImplementedError
+
+    def add_annotation_raw(self, type_id: int, begin: int, end: int) -> int:
+        r"""This function adds an annotation entry with `begin` and `end` index
+        to the sortedlist at index `type_id` of the array which records all
+        sortedlists, return tid for the entry.
+
+        Args:
+            type_id (int): The index of Annotation sortedlist in the array.
+            begin (int): begin index of the entry.
+            end (int): end index of the entry.
+
+        Returns:
+            The tid of the entry.
+        """
+        # We should create the entry list with the format
+        # [entry_type, tid, begin, end, None, ...].
+        # A helper function _new_entry() can be used to generate a entry list
+        # with default fields.
+        # A reference to the entry should be store in both self.elements and
+        # self.entry_dict.
+        raise NotImplementedError
+
+    def set_attr(self, tid: int, attr_name: str, attr_value):
+        r"""This function locates the entry list with tid and sets its
+        attribute `attr_name` with value `attr_value`.
+
+        Args:
+            tid (int): Unique id of the entry.
+            attr_name (str): name of the attribute.
+            attr_value: value of the attribute.
+
+        """
+        # We retrieve the entry list from entry_dict using tid. We get its
+        # entry type. We then locate the index of the attribute using the entry
+        # type, field dictionary and attr_name, and update the attribute.
+
+        raise NotImplementedError
+
+    def get_attr(self, tid: int, attr_name: str):
+        r"""This function locates the entry list with tid and gets the value
+        of `attr_name` of this entry.
+
+        Args:
+            tid (int): Unique id of the entry.
+            attr_name (str): name of the attribute.
+
+        Returns:
+            The value of attr_name for the entry with tid.
+        """
+        # We retrieve the entry list from entry_dict using tid. We get its
+        # entry type. We then locate the index of the attribute using the entry
+        # type, field dictionary and attr_name, and get the attribute.
+
+        raise NotImplementedError
+
+    def delete_entry(self, tid: int):
```

Comment:
there should be a lower-level function of this function, which takes the exact location of the entry and remove it. Using `tid` means you will have to search it first (bisect). This won't be efficient if we already have the bisect results. So there should also be a bisect function:

`delete_entry(tid)` will use `_delete_entry_by_loc(list_id, index_id)` and `bisect`